[AUTOTVM] Tuning with subgraph as granularity

DzAvril opened the pull request:
During performance optimization for platform Xavier which has a Volta GPU in it, we found that tuning by AutoTVM with subgraph as granularity could bring performance improvement.  Because the data type of the subgraph's output may be different from the data type of the subgraph's anchor operator's output, this may change the task from memory-bound to compute-bound or change in reverse.
Let's take the subgraph in the figure below as an example. If we tune with the single convolution the output data type is 'Int32' but if we tune with the subgraph the output data type is 'Int8'. The former's data size is four times the latter one. But in the actual inference, the data type is 'Int8' same as the latter one. So the best config searched by tuning with a single operator maybe not be the best for the subgraph.
![image](https://user-images.githubusercontent.com/18597737/158623733-9f9995d8-ca09-43af-8201-a8109c2ee1f8.png)
We also run an example to verify the theory above.
- We wrote a schedule marked as ```ScheduleA``` for the subgraph above by hardcode and the latency of subgraph inferencing is 104 microseconds. Then we tuned the subgraph with single op as granularity. In the tuning log we found  ```ScheduleA``` and the latency recorded in the measurement result is 329 microseconds. 
- The best schedule from the tuned log in the step above is marked as ```ScheduleB``` and the latency recorded in the measurement result is 237 microseconds. The latency of subgraph inferencing with ```ScheduleB``` is 224 microseconds.

From the example above we can tell AutoTVM would not find ```ScheduleA```,  the obvious better schedule. This means the tuning result is distorted, the distortion would be more obvious if the shape of the output became bigger.
By the way, you may have a question why we don't just tune with AutoSchedule, this is because we want to employ tensor core on Volta GPU to compute matrix multiply and auto-schedule doesn't support tensorize.

This PR proposes a method to tune with subgraph as granularity by AUTOTVM. As we know, tuning with a single operator we need function ```fcompute``` which computes the output tensors of the operator and function ```fschedule``` which schedules the computation. With these two functions, we can build a measurable program running on the device. So the key problem is assigning function ```fcompute``` and function ```fschedule``` for subgraph. In this PR, we use the fucntion  ```fschedule``` of the anchor operator as the subgraph's function ```fschedule```. As for function ```fcompute```, its purpose is getting output tensors of the subgraph. The function ```LowerToTECompute::Lower``` can get the output tensors of the subgraph, so we can use these tensors as output of subgraph's function ```fcompute```. 
The whole process can breakdown into two major phases and it will be clearer reading with code.
1. Task extracting and tuning.
1.1 Select best implementation to compute output  for anchor operator and record implementation name `best_impl_name`
1.2 Lower subgraph to `outputs`
1.3 Create subgraph tuning task name `task_name` with subgraph name and `iotensors` extracted from `outputs`.
1.4 Add subgraph task with `task_name`, `iotensors` and `best_impl_name`.
1.5 Create `workload` for subgraph tuning task with `task_name` and `iotensors`.
1.6 Set function `fcompute`  for subgraph by returning `outputs` .
1.7 Set function `fschedule`  for subgraph by querying table with `best_impl_name`.
1.8 Tune.

2. Building.
2.1 Apply the best history.
2.2 Select best implementation to compute output for anchor operator and record implementation name `best_impl_name`
2.3 Lower subgraph to `outputs`
2.4 Create subgraph tuning task name `task_name` with subgraph name and `iotensors` extracted from `outputs`.
2.5 Create `workload` for subgraph tuning task with `task_name` and `iotensors`.
2.6 Lower schedule with the best config queried by `workload`.
2.7 Codegen.

In our optimization job, employing tuning with subgraph achieved **6.3%** performance improvement.

**But**,  there are two cases in which tuning with subgraph may not get better performance.

Case 1: Anchor operator has more than one implementation. 
We register the subgraph tuning task by `outputs` in step 1.4. No matter how many implementations the anchor operator has, step 1.1 will only pick the implementation with the highest level and return outputs computed by it. So the subgraph tuning tasks may not contain the potential best implementation.

Case 2: Anchor operator's function `fcompute` needs value from config such as code block below. In step 2.2, computing output will call function `_pack_data` and the `cfg` suppose to be the best config of the subgraph. But in step 2.2 we don't know which subgraph the anchor operator belongs to yet, so we cannot get the right config from the best history and fallback to the default one. This may bring great performance regression.
```python
def _pack_data(cfg, data, kernel):
    n, _, ih, iw = get_const_tuple(data.shape)
    oc, ic, kh, kw = get_const_tuple(kernel.shape)
    ic_bn, oc_bn = cfg["tile_ic"].size[-1], cfg["tile_oc"].size[-1]
    ......
```
The cases above seem too hard to resolve in the current implementation. You're welcome to make any suggestions. Thanks!!

DzAvril commented:
cc @FrozenGene.

FrozenGene commented:
@DzAvril Thanks for detail experiment and explaining. As autotvm tuning conv2d, it then will call `traverse_inline` to fuse injective ops in your figure. So the miss information is conv2d's schedule is good but fuse it get worse performance?

DzAvril commented:
@FrozenGene Yes, it only happens when the fused subgraph output data type is different from conv2d's output data type, such as 'Int8' with 'Int32'. The performance bottleneck changes from computation to memory access.

cee1 commented:
> @FrozenGene Yes, it only happens when the fused subgraph output data type is different from conv2d's output data type, such as 'Int8' with 'Int32'. The performance bottleneck changes from computation to memory access.

Let's say a conv and a subgraph containing it,  given a question:
- Which takes more time if the same schedule template and parameters for conv is applied ?

Intuitively, the former takes less time, since it requires lesser computing Ops.

Which is NOT true for qnn:
- a "conv" eats a tensor of dtype=int8/uint8, and outputs a tensor of dtype=int32
- a "subgraph", e.g. "conv + bias + requantize", eats the same input tensor but outputs a tensor of dtype=int8/uint8, thanks to the fused "requantize"

The "subgraph", aka the fused Op, has much lesser "write-back" (dtype=int8/uint8 vs dtype=int32). Which will outperform the single "conv", even the latter one has lesser computing Ops, in our case

Performing AutoTvm on subgraph granularity makes much sense here.

huangxiao2008 commented:
> 
AutoTVM gives the “Best Schedule” according to the tuning result, but a "Less Good Schedule" in the tuning log performs Much Better in inference after doing  traverse_inline.

cee1 commented on python/tvm/autotvm/task/task.py:
```diff
@@ -48,6 +50,105 @@ def _lookup_task(name):
+    args: list of hashable or Tensor
+    """
+
+    def _encode(x):
```
What about use `serialize_args(...)` instead? Which will return a tuple, and then try to hash that tuple (as part of subgraph key)

DzAvril replied:
have added an annotation in code -> get subgraph's workload which is attached to the last op, because the anchor op has attached workload too, so we should get by post order

cee1 commented on python/tvm/autotvm/task/task.py:
```diff
@@ -48,6 +50,105 @@ def _lookup_task(name):
+    for t in io_tensors:
+        # Reject the compute if any of its I/O tensors has dynamic shape.
+        if any([not isinstance(v, int) for v in get_const_tuple(t.shape)]):
+            return []
```
What happens if returns [] here?

Or, may add a FIXME annotation

cee1 commented on python/tvm/autotvm/task/task.py:
```diff
@@ -244,8 +345,12 @@ def __call__(self, *args, **kwargs):
+            arg_bufs = _traverse_to_get_io_tensors(out)
+            s = self.fschedule(out)
+        else:
+            arg_bufs = [out] + self._get_inputs(out)
```
Note: auf_bufs return by "Line 349", has "inputs" first, and then follows "outputs"

cee1 commented on python/tvm/autotvm/task/topi_integration.py:
```diff
@@ -231,14 +239,27 @@ def _decorate(topi_schedule):
-            workload = get_workload(outs, task_name)
+            task_env = TaskExtractEnv.current
+            if GLOBAL_SCOPE.tune_subgraph and task_env is not None and not task_env.tracing:
+                # get workload of subgraph
```
May add an explanation for this `get_workload_post_order(...)`, why necessary?

DzAvril replied:
have added an annotation in code -> get subgraph's workload which is attached to the last op, because the anchor op has attached workload too, so we should get by post order

cee1 commented:
Summary:

For a qnn.conv, a combination of "conv + ... + requantize" will reduce output size by three-quarters(dtype=int32 vs dtype=int8), which is an essential factor for performance tuning, in our case.

Thus, performing AutoTvm on subgraph granularity is important here.

This patch try to add a feature of subgraph-level AutoTvm.
- A `GLOBAL_SCOPE.tune_subgraph` option is introduced.

<br/>

The major change is creating tuning tasks _on subgraph granularity_ in `ScheduleBuilder`, via `register_subgraph_task(...)` and `register_topi_subgraph (...)`

A series of questions need to be handled then:
- To `export_library(...)`, HOWTO hit the tuning log entry of the subgraph?
  - The same subgraph should be generated (tuning time vs export time), i.e. [the same relay passes should be applied](https://github.com/apache/tvm/pull/10650/files#diff-39fc81d81b494a431f3a37d66e6e556c28c71be4b6adcc856c66f2a4cc07bf5dR57)
  - [The same subgraph name should be generated](https://github.com/apache/tvm/pull/10650/files#diff-39fc81d81b494a431f3a37d66e6e556c28c71be4b6adcc856c66f2a4cc07bf5dR215)

- For `register_topi_subgraph (...)`,  which will register a wrapper _fcompute_
  - HOWTO locate the _fcompute_ of the anchor Op? Especially for a _fcompute_ is a wrapper of another _fcompute_ (see [`conv2d_nchw.x86`](https://github.com/apache/tvm/blob/main/python/tvm/relay/op/strategy/x86.py#L134) and [`topi.x86.conv2d_nchw`](https://github.com/apache/tvm/blob/main/python/tvm/topi/x86/conv2d.py#L127) )
  - HOWTO figure out args, specifically, __input args__ of a subgraph, in `TaskTemplate.__call__` 

There are some known limitations of current patch:
1.  In the case of  more than one impl added for the anchor Op
2.  In the case of _fcompute_ accessing config

DzAvril commented:
@cee1 Thanks for the summary. Function `register_topi_subgraph` will also register `fschedule` for subgraph.

DzAvril commented:
There is one more limitation of the current patch. Case 3: 
During task extraction subgraph is lowered by VMCompile, and pass `AlterOpLayout` is disabled, see [code](https://github.com/apache/tvm/blob/main/python/tvm/autotvm/task/relay_integration.py#:~:text=with%20tvm.transform.PassContext(opt_level%3Dopt_level%2C%20disabled_pass%3D%7B%22AlterOpLayout%22%7D)%3A). So during building phash we need to disable pass `AlterOpLayout` too, otherwise the subgraphs generated in the task extracting phase may be different from those generated in the building phase.

DzAvril commented:
In the last few commits, I added a unit test for tuning subgraph. In the test, I set `GLOBAL_SCOPE.tune_subgraph` to `True`, and it seems to lead to many other unit tests failing. The results of the failed tests show that `GLOBAL_SCOPE.tune_subgraph` has been set to `True`. I have no idea why the value set in one test can take effect in another one. So for passing the checks, I removed the unit test. If the community is interested in this patch I could put this test script in the how-to docs.  Attachment is the unit test.
[test_autotvm_subgraph.py.txt](https://github.com/apache/tvm/files/8317096/test_autotvm_subgraph.py.txt)

masahi commented:
@DzAvril For non-trivial PRs like this, we usually ask to send a RFC first at https://github.com/apache/tvm-rfcs. I've just found a good use case for this feature (adding a tunable knob to post-ops fusion). 

cc @tkonolige @comaniac

comaniac commented:
> @DzAvril For non-trivial PRs like this, we usually ask to send a RFC first at https://github.com/apache/tvm-rfcs. I've just found a good use case for this feature (adding a tunable knob to post-ops fusion at
> 
> https://github.com/apache/tvm/blob/3dd1d82efc89730993d54919ce4a61b440474fd5/python/tvm/topi/generic/conv2d.py#L252
> 
> ).
> cc @tkonolige @comaniac

At the first glance this PR includes some core changes (e.g., TE compiler), so I agree to have an RFC first. Meanwhile, I'm supportive to this idea in general. In fact, recently I'm looking for a balance between AutoTVM and auto-scheduler, because I only want certain subgraphs with a fixed anchor op to be tuned with a template (i.e., AutoTVM), and I want to control this behavior directly in the TOPI schedule instead of auto-scheduler's task scheduler. Thus, I could totally understand the motivation and would be happy to discuss, although this PR itself won't serve my personal use cases.

DzAvril commented:
@masahi OK, I will send a Pre-RFC first in the discussion forum.
@comaniac I think we have your use cases too. Our solution is mixing AutoTVM and auto-schedule to tune one model by adding a config `include_op_impls` which lists the certain anchor op with implementation named `conv2d_nchw_int8.cuda` for example. Then AutoTVM only takes charge of certain subgraphs with fixed anchor op listed in `include_op_impls` and auto-schedule takes charge of the others. Combine with the feature proposed in this PR, we can achieve much better performance than tuning with AutoTVM or auto-schedule individually.

masahi commented:
>  I will send a Pre-RFC first in the discussion forum
You don't need to send a pre-rfc, you can go ahead with https://github.com/apache/tvm-rfcs.

I think you just need to copy your PR description over there and summarize the discussion that already happened here.

DzAvril commented:
@masahi @comaniac [RFC](https://github.com/apache/tvm-rfcs/pull/64) is created.
